### PR TITLE
Add support for ambient (top level) functions

### PR DESCRIPTION
**Pull request by winston-riley-zocdoc**
this PR adds support for ambient functions.

### CODE REVIEW

File: typescript/type_literal.go [hunk 1]
```diff
@@ -33,13 +33,17 @@ type PropertySig struct {
 }
 
 func (ps *PropertySig) WriteCode(writer CodeWriter) {
+	ps.writeIdAndTypeAnnotation(writer)
```
Comment (chris-smith-zocdoc) [inline typescript/type_literal.go]: whats this change?

File: typescript/function.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
+package typescript
+
+type Function struct {
```
Comment (chris-smith-zocdoc) [inline typescript/function.go]: want to add support for arrow functions? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

File: typescript/function.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	Name            string
+	body            BlockDeclaration
+	params          []ParamDeclaration
+	returnType      Writable
+	typeConstructor Writable
+	exported        bool
```
Comment (chris-smith-zocdoc) [inline typescript/function.go]: add support for async keyword

File: typescript/function.go (hunk 2 above)
Comment (winston-riley-zocdoc) [reply typescript/function.go]: good call. ill look at it

File: typescript/type_literal.go (hunk 1 above)
Comment (winston-riley-zocdoc) [reply typescript/type_literal.go]: It was some intermediate code that I was using that i forgot to cleanup. its gone now.

File: typescript/function.go (hunk 3 above)
Comment (winston-riley-zocdoc) [reply typescript/function.go]: done

File: typescript/function.go (hunk 2 above)
Comment (winston-riley-zocdoc) [reply typescript/function.go]: ok, added support for arrow, give it another look. 

This is currently only supporting the most general form:
```
(param1, paramN) => {
  statements
}
```

**Review by christopher-taormina-zocdoc — APPROVED** [review]

**Comment by winston-riley-zocdoc** [comment]
@christopher-taormina-zocdoc, I realized I had not implemented a sufficient public api. Sorry for the oversight and the resulting re-review request. 

The gist of the [change](https://github.com/Zocdoc/gopoetry/pull/18/commits/f1db356867ba6cdecb7717ce0998c80e92654dee) is that I re-wrote the tests to use the public api, where before I was constructing the values with private fields.

**Review by christopher-taormina-zocdoc — APPROVED** [review]
Oh man I didn't even notice before, i've been away from this codebase for too long

File: typescript/function.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
+package typescript
+
+// FunctionDecl represents a top level declaration.
+type FunctionDecl struct {
+	Name     string
+	Exported bool
```
Comment (christopher-taormina-zocdoc) [inline typescript/function.go]: why does Exported need to be public?

Comment (winston-riley-zocdoc) [reply typescript/function.go]: nope, good catch. Ill make it private

**Pull request merged**